crates/aptos-protos/proto/aptos/transaction/v1/transaction.proto

```diff
@@ -0,0 +1,415 @@
+  string key_type = 2;
+  string value = 3;
+  string value_type = 4;
+}
+
+message WriteTableItem {
+  bytes state_key_hash = 1;
+  string handle = 2;
+  string key = 3;
+  WriteTableData data = 4;
+}
+
+message TransactionPayload {
+  enum Type {
+    ENTRY_FUNCTION_PAYLOAD = 0;
+    SCRIPT_PAYLOAD = 1;
+    MODULE_BUNDLE_PAYLOAD = 2;
+  }
+
+  Type type = 1;
+
+  oneof payload {
+    EntryFunctionPayload entry_function_payload = 2;
+    ScriptPayload script_payload = 3;
+    ModuleBundlePayload module_bundle_payload = 4;
+    WriteSetPayload write_set_payload = 5;
+  }
+}
+
+message EntryFunctionPayload {
+  EntryFunctionId function = 1;
+  repeated MoveType type_arguments = 2;
+  repeated string arguments = 3;
+}
+
+message MoveScriptBytecode {
+  bytes bytecode = 1;
+  MoveFunction abi = 2;
+}
+
+message ScriptPayload {
+  MoveScriptBytecode code = 1;
+  repeated MoveType type_arguments = 2;
+  repeated string arguments = 3;
+}
+
+message ModuleBundlePayload {
+  repeated MoveModuleBytecode modules = 1;
+}
+
+message MoveModuleBytecode {
+  bytes bytecode = 1;
+  MoveModule abi = 2;
+}
+
+message MoveModule {
+  string address = 1;
+  string name = 2;
+  repeated MoveModuleId friends = 3;
+  repeated MoveFunction exposed_functions = 4;
+  repeated MoveStruct structs = 5;
+}
+message MoveFunction {
+  enum Visibility {
+    PRIVATE = 0;
+    PUBLIC = 1;
+    FRIEND = 2;
+  }
+  string name = 1;
+  MoveFunction.Visibility visibility = 2;
+  bool is_entry = 3;
+  repeated MoveFunctionGenericTypeParam generic_type_params = 4;
+  repeated MoveType params = 5;
+  repeated MoveType return = 6;
+}
+
+message MoveStruct {
+  string name = 1;
+  bool is_native = 2;
+  repeated MoveAbility abilities = 3;
+  repeated MoveStructGenericTypeParam generic_type_params = 4;
+  repeated MoveStructField fields = 5;
+}
+
+message MoveStructGenericTypeParam{
+  repeated MoveAbility constraints = 1;
+  bool is_phantom = 2;
+}
+
+message MoveStructField {
+  string name = 1;
+  MoveType type = 2;
+}
+
+message MoveFunctionGenericTypeParam {
+  repeated MoveAbility constraints = 1;
+}
+
+// Next id: 12
+enum MoveTypes {
+  Bool = 0;
+  U8 = 1;
+  U16 = 11;
+  U32 = 12;
+  U64 = 2;
+  U128 = 3;
+  U256 = 13;
+  Address = 4;
+  Signer = 5;
+  Vector = 6; // { items: Box<MoveType> },
+  Struct = 7; //(MoveStructTag),
+  GenericTypeParam = 8; // { index: u16 },
+  Reference = 9; // { mutable: bool, to: Box<MoveType> },
+  Unparsable = 10; //(String),
+}
+
+message MoveType{
+
+  message ReferenceType {
+    bool mutable = 1;
+    MoveType to = 2;
+  }
+
+  MoveTypes type = 1;
+  oneof content {
+    MoveType vector = 3;
+    MoveStructTag struct = 4;
+    uint32 generic_type_param_index = 5;
+    ReferenceType reference = 6;
+    string unparsable = 7;
+  }
+}
+
+enum MoveAbility {
+  COPY = 0;
+  DROP = 1;
+  STORE = 2;
+  KEY = 3;
+}
+
+message WriteSetPayload {
+  WriteSet write_set = 1;
+}
+
+message EntryFunctionId {
+  MoveModuleId module = 1;
+  string name = 2;
```

Comment:
Why are we exposing all the internal details of the Move VM in this proto file? This sounds like a maintenance nightmare to me. What happens for example if we add a new ability in Move, do we need to go here and fix this code?

Reply:
Open to suggestions here. This version mimics the API types. Maybe I'm missing something but at this moment I'm thinking that if we add an incompatible BCS change, we'll need to upgrade everything that works with BCS, e.g. API, SDK, and I guess indexer here. Is there another way?

Reply:
I think the ideal situation would be that this is generated from the rust code, but I don't suspect there's a rust to proto converter?

Reply:
I think the ideal situation would be that this is generated from the rust code, but I don't suspect there's a rust to proto converter?